File: svc-position/src/main/java/com/beancounter/position/accumulation/TrnBehaviourFactory.java
package com.beancounter.position.accumulation;

import com.beancounter.common.model.TrnType;
import java.util.HashMap;
import java.util.Map;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class TrnBehaviourFactory {

  private final Map<TrnType, AccumulationStrategy> trnBehaviours = new HashMap<>();

  @Autowired(required = false)
  void setBuyBehaviour(BuyBehaviour buyBehaviour) {
    trnBehaviours.put(TrnType.BUY, buyBehaviour);
  }

  @Autowired(required = false)
  void setSellBehaviour(SellBehaviour sellBehaviour) {
    trnBehaviours.put(TrnType.SELL, sellBehaviour);
  }

  @Autowired(required = false)
  void setSplitBehaviour(SplitBehaviour splitBeahviour) {
    trnBehaviours.put(TrnType.SPLIT, splitBeahviour);
  }

  @Autowired(required = false)
  void setDividendBehaviour(DividendBehaviour dividendBeahviour) {
    trnBehaviours.put(TrnType.DIVI, dividendBeahviour);
  }

  public AccumulationStrategy get(TrnType trnType) {
    return trnBehaviours.get(trnType);
  }
}

File: svc-position/src/main/java/com/beancounter/position/accumulation/TrnBehaviourFactory.kt
package com.beancounter.position.accumulation

import com.beancounter.common.model.TrnType
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.EnumMap

@Service
class TrnBehaviourFactory {
    private val trnBehaviours: MutableMap<TrnType, AccumulationStrategy> = EnumMap(TrnType::class.java)

    @Autowired(required = false)
    fun setBuyBehaviour(buyBehaviour: BuyBehaviour) {
        trnBehaviours[TrnType.BUY] = buyBehaviour
    }

    @Autowired(required = false)
    fun setSellBehaviour(sellBehaviour: SellBehaviour) {
        trnBehaviours[TrnType.SELL] = sellBehaviour
    }

    @Autowired(required = false)
    fun setSplitBehaviour(splitBeahviour: SplitBehaviour) {
        trnBehaviours[TrnType.SPLIT] = splitBeahviour
    }

    @Autowired(required = false)
    fun setDividendBehaviour(dividendBeahviour: DividendBehaviour) {
        trnBehaviours[TrnType.DIVI] = dividendBeahviour
    }

    operator fun get(trnType: TrnType?): AccumulationStrategy? {
        return trnBehaviours[trnType]
    }
}
